src/multiformats/address.cljc

```diff
@@ -2,29 +2,232 @@
+     dst))
+  ([src offset]
+   (copy-slice src offset (- (alength ^bytes src) offset))))
+
+
+(defn- decode-value
+  "decode value using protocol `codec` in `data` starting at `offset`"
+  [codec ^bytes data offset]
+  (if-let [fixed-len (codec/fixed-byte-length codec)]
+    (let [value-bytes (copy-slice data offset fixed-len)]
+      [(codec/bytes->str codec value-bytes) fixed-len])
+    ;; varint encoding
+    (let [[value-len num-read] (varint/read-bytes data offset)
+          value-bytes (copy-slice data (+ offset num-read) value-len)]
+      [(codec/bytes->str codec value-bytes) (+ num-read value-len)])))
+
+
+(defn- protocol-value-seq
+  "lazy sequence of `[protocol-key value]`, where
+  `value` is a string representation of the address protocol
+   value of `nil` if protocol has no value"
+  ([^bytes data]
+   (protocol-value-seq data 0))
+  ([^bytes data offset]
+   (when (< offset (alength data))
+       (lazy-seq
+        (let [[code code-len] (varint/read-bytes data offset)
+              protocol-key (get code->protocol code)
+              {:keys [codec]} (ensure-protocol-attrs protocol-key)
+              val-start (+ offset code-len)
+              [value num-read] (decode-value codec data val-start)]
+          (cons [protocol-key value]
+                (protocol-value-seq data (+ offset code-len num-read))))))))
+
+
+(defn- concat-arrs [& arrs]
+  (let [total-len (reduce + (map alength arrs))
+        dst (b/byte-array total-len)]
+    (loop [arrs arrs offset 0]
+      (when-let [^bytes src (first arrs)]
+        (b/copy src 0 dst offset (alength src))
+        (recur (rest arrs) (+ offset (alength src)))))
+    dst))
+
+(defn- entry-bytes
+  "build byte array representation for protocol/value pair,
+   note that `value` may be nil for no-value protocols.
+
+   entry can either be a single keyword for a no-value protocol, or a
+   pair of [protocol-key value]. You can also pass in [key nil] for a protocol
+   not expecting a value"
+  [entry]
+  (let [protocol-key (if (keyword? entry) entry (first entry))
+        value (when (coll? entry) (second entry))
+        {:keys [code, codec]} (ensure-protocol-attrs protocol-key)
+        code-bytes (varint/encode code)
+        val-bytes (if (nil? value)
+                    (b/byte-array 0)
+                    (codec/str->bytes codec value))
+        val-len-bytes (if (codec/fixed-byte-length codec)
+                        (b/byte-array 0)
+                        (varint/encode (alength ^bytes val-bytes)))]
+    (concat-arrs code-bytes val-len-bytes val-bytes)))
+
+(deftype Address
+  [^bytes _data _meta ^:unsynchronized-mutable _hash]
+
+  #?(:clj Seqable :cljs ISeqable)
+  (#?(:clj seq :cljs -seq) [this] (protocol-value-seq _data))
+
+  #?(:clj IPersistentCollection)
+
+  #?(:cljs ICounted)
+  (#?(:clj count :cljs -count)
+    [this] (count (seq this)))
+
+  #?(:cljs IEmptyableCollection)
+  (#?(:clj empty :cljs -empty)
+    [this] (Address. (b/byte-array 0) nil 0))
```

Comment:
I believe that standard behavior is to preserve metadata on emptied collections:
```clojure
=> (-> {} (with-meta {:foo 123}) empty meta)
{:foo 123}
```